File: core/java/src/main/java/hara/lang/protocol/IPair.java
package hara.lang.protocol;

import hara.lang.base.Ex;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

import java.util.Map;

@HaraProtocolBinding(namespace = "std.protocol.ipair", name = "IPair")
public interface IPair<K, V> extends Map.Entry<K, V> {
  @HaraMethod(value = "key", arity = 1)
  @Override
  K getKey();

  @HaraMethod(value = "value", arity = 1)
  @Override
  V getValue();

  @Override
  default V setValue(V value) {
    throw new Ex.Unsupported();
  }
}

File: core/rust/src/lang/protocol/ipair.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.ipair", name = "IPair")]
pub trait IPair<K, V> {
    #[hara_method(value = "key", arity = 1)]
    fn key(&self) -> &K;
    #[hara_method(value = "value", arity = 1)]
    fn value(&self) -> &V;
}
